<a href="https://colab.research.google.com/github/gabrielawad/talleresGoogleColab/blob/main/Taller_15_CVXPY_solvers_Branch_and_Bound.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## «*Hay personas que tienen la memoria muy llena, pero el juicio muy vacío y hueco*».
### [Miguel de Montaigne](https://es.wikipedia.org/wiki/Michel_de_Montaigne)

# Taller 15 CVXPY solvers y Branch and Bound



# Características generales de CVXPY

CVXPY  es un lenguaje de modelado para [optimización de problemas convexos](https://web.stanford.edu/~boyd/cvxbook/), que funciona como una librería de Python.

Su principal ventaja es que permite la formulación de problemas por medio de expresiones algebraicas o matriciales. CVXPY automáticamente transforma el problema a su forma estándar, lo resuelve por medio de un  [solvers](https://en.wikipedia.org/wiki/Solver), y entrega los resultados.

Es decir, simplifica el proceso de optimización computacional ya que es capaz de resolver el problema sin necesidad de convertirlo a las formas estándares requeridas por los [solvers](https://en.wikipedia.org/wiki/Solver).

## Página oficial
La información oficial de CVXPY puede ser consultada en [www.cvxpy.org](http://www.cvxpy.org)

# Estados de un problema

CVXPY identifica los siguientes estados en la solución de un problema:
* OPTIMAL  (se encontró al menos una solución óptima)
* INFEASIBLE (no se encontraron soluciones factibles)
* UNBOUNDED (se encontró una solución indeterminada)
* OPTIMAL_INACCURATE (se encontró al menos una solución óptima, **pero el nivel de precisión es inferior al deseado**)
* INFEASIBLE_INACCURATE (no se encontraron soluciones factibles **y el nivel de precisión es inferior al deseado**)
* UNBOUNDED_INACCURATE  (se encontró una solución indeterminada **y el nivel de precisión es inferior al deseado**)

En los casos diferentes a estados **OPTIMAL** se sugiere revisar la formulación del problema y los valores de los parámetros.

En los casos de estados **INACCURATE** de la solución se recomienda ensayar a resolver el problema utilizando otro solver.


# Gestión de solvers

CVXPY utiliza open source solvers como [OSQP](https://osqp.org), [ECOS](https://www.embotech.com/ECOS), [ECOS_BB](https://github.com/embotech/ecos#mixed-integer-socps-ecos_bb) y [SCS](https://github.com/cvxgrp/scs). Si se desea utilizar otros solvers ([CVXOPT](http://cvxopt.org), [GLPK](https://www.gnu.org/software/glpk/),  [CBC](https://projects.coin-or.org/Cbc),  [MOSEK](https://www.mosek.com),  [GUROBI](http://www.gurobi.com)) se debe instalar cada uno de ellos de manera adicional.
El código que se presenta a continuación  resuelve el mismo ejercicio utilizando cuatro solvers diferentes. Analice los valores obtenidos tanto para la solución óptima como para las variables de decisión.

In [ ]:
# Este programa presenta la solución de un ejercicio de optimización
# utilizando cuatro solvers

# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo
obj = cvx.Minimize(3*x + 8*y)

# Establecer las restricciones
constraints = [
x + y >= 50,
x - y <= 20,
x >= 0,
y >= 0
]

# Configurar el problema
prob_ej = cvx.Problem(obj, constraints)

# Obtener el valor óptimo, actualizar el estado del problema y los valores
# de todas las variables del problema
prob_ej.solve()

print ("Soluciones sin especificar el solver")

# Mostrar el estado de la solución sin especificar el solver
print("Estado de la solución sin especificar el solver:", prob_ej.status)

# Mostrar el valor óptimo de la función objetivo sin especificar el solver
print("Valor óptimo de la función objetivo sin especificar el solver:",
      prob_ej.value)

# Mostrar los valores óptimos de las variables de decisión sin especificar
# el solver
print("Valores óptimos de las variables de decisión sin especificar el solver:",
      x.value, y.value, "\n")

# Soluciones con OSQP
print ("Soluciones con OSQP")

# Resolver con OSQP
prob_ej.solve(solver=cvx.OSQP)

# Mostrar el estado de la solución con OSQP
print("Estado de la solución con OSQP:", prob_ej.status)

print("Valor óptimo de la función objetivo con OSQP:", prob_ej.value)

# Mostrar los valores óptimos de las variables de decisión con OSQP
print("Valores óptimos de las variables de decisión con OSQP:",
      x.value, y.value, "\n")

# Soluciones con ECOS
print ("Soluciones con ECOS")

# Resolver con ECOS
prob_ej.solve(solver=cvx.ECOS)

# Mostrar el estado de la solución con ECOS
print("Estado de la solución con ECOS:", prob_ej.status)

print("Valor óptimo de la función objetivo con ECOS:", prob_ej.value)

# Mostrar los valores óptimos de las variables de decisión con ECOS
print("Valores óptimos de las variables de decisión con ECOS:",
      x.value, y.value, "\n")

# Soluciones con ECOS_BB
print ("Soluciones con ECOS_BB")

# Resolver con ECOS_BB
prob_ej.solve(solver=cvx.ECOS_BB)

# Mostrar el estado de la solución  con ECOS_BB
print("Estado de la solución con ECOS_BB:", prob_ej.status)

print("Valor óptimo de la función objetivo con ECOS_BB:", prob_ej.value)

# Mostrar los valores óptimos de las variables de decisión con ECOS_BB
print("Valores óptimos de las variables de decisión con ECOS_BB:",
      x.value, y.value, "\n")

# Soluciones con SCS
print ("Soluciones con SCS")

# Resolver con SCS
prob_ej.solve(solver=cvx.SCS)

# Mostrar el estado de la solución  con SCS
print("Estado de la solución con SCS:", prob_ej.status)

print("Valor óptimo de la función objetivo con SCS:", prob_ej.value)

# Mostrar los valores óptimos de las variables de decisión con SCS
print("Valores óptimos de las variables de decisión con SCS:",
      x.value, y.value, "\n")

Soluciones sin especificar el solver
Estado de la solución sin especificar el solver: optimal
Valor óptimo de la función objetivo sin especificar el solver: 224.99999999004348
Valores óptimos de las variables de decisión sin especificar el solver: 34.99999999851025 14.999999999314094 

Soluciones con OSQP
Estado de la solución con OSQP: optimal
Valor óptimo de la función objetivo con OSQP: 225.0
Valores óptimos de las variables de decisión con OSQP: 35.0 15.0 

Soluciones con ECOS
Estado de la solución con ECOS: optimal
Valor óptimo de la función objetivo con ECOS: 224.9999996816037
Valores óptimos de las variables de decisión con ECOS: 35.00000000022704 14.999999960115321 

Soluciones con ECOS_BB
Estado de la solución con ECOS_BB: optimal
Valor óptimo de la función objetivo con ECOS_BB: 224.9999996816037
Valores óptimos de las variables de decisión con ECOS_BB: 35.00000000022704 14.999999960115321 

Soluciones con SCS
Estado de la solución con SCS: optimal
Valor óptimo de la función o

# Branch and Bound

Ejemplo de la solución  paso a paso de un problema de programación entera utilizando la heurística Branch and Bound.

## Problema inicial: problema00

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones00 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob00 = cvx.Problem(obj, restricciones00)

# Obtener el valor óptimo
prob00.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob00.status)
print("Valor máximo:", "%.2f"% prob00.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.25
Valor óptimo de x: 3.75
Valor óptimo de y: 2.25


## subproblema01

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45
   
   **x <= 3**

   x   > = 0      

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones01 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x <= 3,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob01 = cvx.Problem(obj, restricciones01)

# Obtener el valor óptimo
prob01.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob01.status)
print("Valor máximo:", "%.2f"% prob01.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 39.00
Valor óptimo de x: 3.00
Valor óptimo de y: 3.00


## subproblema02

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones02 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob02 = cvx.Problem(obj, restricciones02)

# Obtener el valor óptimo
prob02.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob02.status)
print("Valor máximo:", "%.2f"% prob02.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.80


## subproblema03

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

  **y >= 2**

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones03 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
y >= 2

# Restricciones de no negatividad

]

# Configurar el problema
prob03 = cvx.Problem(obj, restricciones03)

# Obtener el valor óptimo
prob03.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob03.status)
print("Valor máximo:", "%.2f"% prob03.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: infeasible
Valor máximo: -inf


TypeError: ignored

## subproblema04

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones04 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob04 = cvx.Problem(obj, restricciones04)

# Obtener el valor óptimo
prob04.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob04.status)
print("Valor máximo:", "%.2f"% prob04.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.56
Valor óptimo de x: 4.44
Valor óptimo de y: 1.00


## subproblema05

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 5**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones05 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 5,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob05 = cvx.Problem(obj, restricciones05)

# Obtener el valor óptimo
prob05.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob05.status)
print("Valor máximo:", "%.2f"% prob05.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00


## subproblema06

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **x <= 4**
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear la variable a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones06 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
x <=4,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob06 = cvx.Problem(obj, restricciones06)

# Obtener el valor óptimo
prob06.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob06.status)
print("Valor máximo:", "%.2f"% prob06.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 37.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.00


# Resumen del Branch and Bound
## Problema inicial: problema00
Restricciones: x>=0; y>=0     

Resultados: valor óptimo: 41.25; x.value = 3.75; y.value = 2.25

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0     

Resultados: valor óptimo: 39.00; **x.value = 3.00**; **y.value = 3.00**

## subproblema02
Restricciones: **x>=4**; y>=0     

Resultados: valor óptimo: 41.00; x.value = 4.00; y.value = 1.80

## subproblema03
Restricciones: **x>=4**; **y >=2**     

Resultados: valor óptimo: infeasible; x.value = ???; y.value = ???

## subproblema04
Restricciones: **x>=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 40.56; x.value = 4.44; y.value = 1.00

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: valor óptimo: 40.00; **x.value = 5.00; y.value = 0.00**

## subproblema06
Restricciones: **x>=4**; **x<=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 37.00; **x.value = 4.0; y.value = 1.0**

# Análisis de los resultados del Branch and Bound

Los subproblemas 01, 05 y 06 tienen una solución óptima factible (soluciones enteras).

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0

Resultados: **valor óptimo: 39.00; x.value = 3.00; y.value = 3.00**

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: **valor óptimo: 40.00; x.value = 4.0; y.value = 1.0**

## subproblema06
Restricciones: x>=4; x<=4; y>=0; y <=1

Resultados: **valor óptimo: 37.00; x.value = 4.0; y.value = 1.0**

Se selecciona como solución optima la del subproblema 05, ya que satisface las restricciones de enteros y genera el mayor valor óptimo.


# Problema inicial: problema00 resuelto directamente por CVXPY

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [ ]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar y establecerlas como enteros
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones_int = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob_int = cvx.Problem(obj, restricciones_int)

# Obtener el valor óptimo
prob_int.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob_int.status)
print("Valor máximo:", "%.2f"% prob_int.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value, "\n")

# Soluciones con ECOS_BB
print ("Soluciones con ECOS_BB")

# Resolver con ECOS_bb
prob_int.solve(solver=cvx.ECOS_BB, mi_max_iters = 1000,
               mi_abs_eps= 1e-6, mi_rel_eps = 1e-3, verbose = True)

# prob_int.solve(solver=cvx.ECOS_BB, mi_max_iters = 1000,
#               mi_abs_eps= 1e-6, mi_rel_eps = 1e-3)

# Mostrar el estado de la solución  con ECOS_BB
print("Estado de la solución con ECOS_BB:", prob_int.status)

print("Valor óptimo de la función objetivo con ECOS_BB:", prob_int.value)

# Mostrar los valores óptimos de las variables de decisión con ECOS_BB
print("Valores óptimos de las variables de decisión con ECOS_BB:",
      x.value, y.value, "\n")

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00 

Soluciones con ECOS_BB
                                     CVXPY                                     
                                     v1.3.2                                    
(CVXPY) Oct 03 03:28:21 PM: Your problem has 2 variables, 4 constraints, and 0 parameters.
(CVXPY) Oct 03 03:28:21 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Oct 03 03:28:21 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Oct 03 03:28:21 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Oct 03 03:28:21 PM: Compiling pro

# Problema inicial: problema00 resuelto con MOSEK

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**

## Instalar MOSEK

In [ ]:
# Instalar MOSEK en Google Colab
!pip install mosek

## Verificar la instalación de MOSEK

In [ ]:
# Importar las librerías requeridas
import cvxpy as cvx
import mosek

# Verificar los solvers disponibles
print(cvx.installed_solvers())

['CVXOPT', 'ECOS', 'ECOS_BB', 'GLPK', 'GLPK_MI', 'MOSEK', 'OSQP', 'SCIPY', 'SCS']


## Aviso importante

Antes de ejecutar cvxpy con el solver MOSEK es necesario subir el archivo con la licencia de MOSEK (mosek.lic) para crear la ruta /root/mosek/mosek.lic

**Este paso hay que repetirlo cada vez que se vaya a utilizar MOSEK porque desaparece cuando se reinicia el entorno de ejecución**.

En el siguiente enlace puede solicitar una licencia académica **gratuita** de MOSEK: [MOSEK Academic Licenses](https://www.mosek.com/products/academic-licenses/)

## Solución del problema00 con MOSEK

In [ ]:
# Importar las librerías que se van a utilizar
import cvxpy as cvx
from mosek.fusion import *

# Crear las variables a optimizar y establecerlas como enteros
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones_msk = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob_msk = cvx.Problem(obj, restricciones_msk)

# Obtener el valor óptimo
prob_msk.solve()

# Resolver con MOSEK
prob_msk.solve(solver=cvx.MOSEK)
#prob_msk.solve(solver=cvx.MOSEK, verbose = True)

# Mostrar el estado de la solución  con MOSEK
print("Estado de la solución con MOSEK:", prob_msk.status)

print("Valor óptimo de la función objetivo con MOSEK:", prob_msk.value)

# Mostrar los valores óptimos de las variables de decisión con MOSEK
print("Valores óptimos de las variables de decisión con ")
print("Valor óptimo de x con MOSEK:", x.value)
print("Valor óptimo de y con MOSEK:", y.value)

Estado de la solución con MOSEK: optimal
Valor óptimo de la función objetivo con MOSEK: 40.0
Valores óptimos de las variables de decisión con 
Valor óptimo de x con MOSEK: 5.0
Valor óptimo de y con MOSEK: 0.0


# Ejercicios

## 00.
Resuelva con el método **Branch and Bound**, utilizando CVXPY, el ejemplo 9.2-1 del libro de (TAHA, 2012) página 336.

## 01.
Resuelva directamente, utilizando CVXPY, el ejemplo 9.2-1 del libro de (TAHA, 2012) página 336.

## 02.
Resuelva con MOSEK, utilizando CVXPY, el ejemplo 9.2-1 del libro de (TAHA, 2012) página 336.

## 03.
Resuelva con el método **Branch and Bound**, utilizando CVXPY, **uno** de los cinco numerales del ejercicio 2 del conjunto de problemas 9.2a del libro de (TAHA, 2012) página 342.

## 04.
Resuelva directamente, utilizando CVXPY, **uno** de los cinco numerales del ejercicio 2 del conjunto de problemas 9.2a del libro de (TAHA, 2012) página 342.

Sugerencia: utilice el mismo numeral utilizado en el ejercicio 03.

## 05.
Resuelva con MOSEK, utilizando CVXPY, **uno** de los cinco numerales del ejercicio 2 del conjunto de problemas 9.2a del libro de (TAHA, 2012) página 342.

Sugerencia: utilice el mismo numeral utilizado en el ejercicio 03.